### PR TITLE
feat(dashmate): re-use ZeroSSL private key

### PR DESCRIPTION
**Pull request by shumkov**
## Issue being fixed or feature implemented

To simplify infrastructure, we want to use the same private key for ZeroSSL certificates for a given node. Current renewal logic creates and new private key and certificate request every time you run it

## What was done?

- Store and reuse certificate request file for renewals
- Validate ZeroSSL configuration and files
- Download certificate file only if necessary
- Remove unsafe retries from the `downloadCertificate` function

## How Has This Been Tested?

- Manually running the `dashmate ssl obtain` command
- Running with https://github.com/dashpay/dash-network-deploy/pull/516

## Breaking Changes

None

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js [hunk 1]
```diff
@@ -33,56 +35,135 @@ function obtainZeroSSLCertificateTaskFactory(
...
+          ctx.isCrtFilePresent = fs.existsSync(csrFilePath);
+
+          ctx.isPrivateKeyFilePresent = fs.existsSync(privateKeyFilePath);
+
+          ctx.isBundleFilePresent = fs.existsSync(bundleFilePath);
+
+          // This function will throw an error if certificate with specified ID is not present
+          const certificate = await getCertificate(apiKey, certificateId);
+
+          // If certificate exists but private key is not, then we can't setup TLS connection
```
Comment (strophy) [inline packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js:78]:
```suggestion
          // If certificate exists but private key does not, then we can't setup TLS connection
```

File: packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js [hunk 2]
```diff
@@ -33,56 +35,135 @@ function obtainZeroSSLCertificateTaskFactory(
...
+          // In this case we need to regenerate certificate or put back this private key
+          if (!ctx.isPrivateKeyFilePresent) {
+            throw new Error(`Certificate private key file is not found in ${privateKeyFilePath}.\n`
```
Comment (strophy) [inline packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js:81]:
```suggestion
            throw new Error(`Certificate private key file not found in ${privateKeyFilePath}.\n`
```

File: packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js [hunk 3]
```diff
@@ -33,56 +35,135 @@ function obtainZeroSSLCertificateTaskFactory(
...
+              + 'Please regenerate the certificate using the the obtain'
+              + ' command with --force flag and revoke previous certificate in ZeroSSL'
+              + ' dashboard');
```
Comment (strophy) [inline packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js:84]:
```suggestion
              + 'Please regenerate the certificate using the the obtain'
              + ' command with the --force flag, and revoke the previous certificate in'
              + ' the ZeroSSL dashboard');
```

File: packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js [hunk 4]
```diff
@@ -33,56 +35,135 @@ function obtainZeroSSLCertificateTaskFactory(
...
+          }
+
+          // We need to make sure that external IP and certificate IP match
+          if (certificate.common_name !== externalIp) {
+            throw new Error(`Certificate IPe ${certificate.common_name} does not match external IP ${externalIp}.\n`
+            + 'Please change external IP or regenerate the certificate using the obtain'
+            + ' command with --force flag and revoke previous certificate in ZeroSSL'
+            + ' dashboard');
```
Comment (strophy) [inline packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js:92]:
```suggestion
            + 'Please change the external IP in config or regenerate the certificate '
            + ' using the obtain command with the --force flag, and revoke the previous'
            + ' certificate in the ZeroSSL dashboard');
```

File: packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js [hunk 5]
```diff
@@ -33,56 +35,135 @@ function obtainZeroSSLCertificateTaskFactory(
...
 
-                // eslint-disable-next-line no-param-reassign
-                task.output = `Certificate already exists and expires at ${ctx.certificate.expires}`;
-              }
-            // Certificate is going to expire soon, we need to obtain a new one
+              // We need to download new certificate bundle
+              ctx.isBundleFilePresent = false;
+
+              // eslint-disable-next-line no-param-reassign
+              task.output = 'Certificate is already created, but not validated yet.';
```
Comment (strophy) [inline packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js:113]:
```suggestion
              task.output = 'Certificate was already created, but not validated yet.';
```

File: packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js [hunk 6]
```diff
@@ -33,56 +35,135 @@ function obtainZeroSSLCertificateTaskFactory(
...
             } else {
+              // Certificate is not valid, so we need to re-create it
+
+              // We need to download certificate bundle
+              ctx.isBundleFilePresent = false;
+
+              if (!ctx.isCrtFilePresent) {
+                throw new Error(`Certificate request file is not found in ${csrFilePath}.\n`
```
Comment (strophy) [inline packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js:121]:
```suggestion
                throw new Error(`Certificate request file not found in ${csrFilePath}.\n`
```

File: packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js [hunk 7]
```diff
@@ -33,56 +35,135 @@ function obtainZeroSSLCertificateTaskFactory(
...
+                  + 'To create a new certificate please use the obtain'
+                  + ' command with --force flag and revoke previous certificate in ZeroSSL'
+                  + ' dashboard');
```
Comment (strophy) [inline packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js:124]:
```suggestion
                  + 'To create a new certificate, please use the obtain'
                  + ' command with the --force flag and revoke the previous certificate'
                  + ' in the ZeroSSL dashboard');
```

File: packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js [hunk 8]
```diff
@@ -33,56 +35,135 @@ function obtainZeroSSLCertificateTaskFactory(
...
+            // Certificate is going to expire soon, we need to obtain a new one
+
+            // We need to download new certificate bundle
+            ctx.isBundleFilePresent = false;
+
+            if (!ctx.isCrtFilePresent) {
+              throw new Error(`Certificate request file is not found in ${csrFilePath}.\n`
+                + 'To renew certificate please use the obtain'
+                + ' command with --force flag and revoke previous certificate in ZeroSSL'
+                + ' dashboard');
```
Comment (strophy) [inline packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js:142]:
```suggestion
                + 'To renew certificate please use the obtain'
                + ' command with the --force flag, and revoke the previous certificate in'
                + ' the ZeroSSL dashboard');
```

File: packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js [hunk 9]
```diff
@@ -33,56 +35,135 @@ function obtainZeroSSLCertificateTaskFactory(
...
+              task.output = 'Certificate is not valid. Create a new one';
+            }
+          } else {
+            // Certificate is going to expire soon, we need to obtain a new one
+
+            // We need to download new certificate bundle
+            ctx.isBundleFilePresent = false;
+
+            if (!ctx.isCrtFilePresent) {
+              throw new Error(`Certificate request file is not found in ${csrFilePath}.\n`
```
Comment (strophy) [inline packages/dashmate/src/listr/tasks/ssl/zerossl/obtainZeroSSLCertificateTaskFactory.js:139]:
```suggestion
              throw new Error(`Certificate request file not found in ${csrFilePath}.\n`
```

**Review by strophy — CHANGES_REQUESTED** [review]

**Review by strophy — DISMISSED** [review]

**Review by jawid-h — DISMISSED** [review]

**Review by strophy — APPROVED** [review]
:rocket:

**Pull request merged**